builtin/logical/pki/crl_util.go

```diff
@@ -0,0 +1,153 @@
+package pki
+
+import (
+	"crypto/rand"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"fmt"
+	"time"
+
+	"github.com/hashicorp/vault/logical"
+)
+
+type revocationInfo struct {
+	CertificateBytes []byte `json:"certificate_bytes"`
+	RevocationTime   int64  `json:"unix_time"`
+}
+
+func revokeCert(req *logical.Request, serial string) (*logical.Response, error) {
+	certEntry, userErr, intErr := fetchCertBySerial(req, "revoked/", serial)
+	if certEntry != nil {
+		return nil, nil
+	}
+
+	certEntry, userErr, intErr = fetchCertBySerial(req, "certs/", serial)
+	switch {
+	case userErr != nil:
+		return logical.ErrorResponse(userErr.Error()), nil
+	case intErr != nil:
+		return nil, intErr
+	}
+
+	// Possible TODO: use some kind of transaction log in case of an
+	// error anywhere along here (so we've validated that we got a
+	// value back, but want to make sure it not only is deleted from
+	// certs/ but also shows up in revoked/ and a CRL is generated)
+	err := req.Storage.Delete("certs/" + serial)
```

Comment:
Yeah. In a moment I will push some changes that should make it robust -- it will delay deleting until the end, and if it finds a cert in revoked/ it will still check to see if it is in certs/. If so, it will rebuild the CRL and attempt to delete in certs/ again.